### PR TITLE
UCP/AM: Added eager reply.

### PR DESCRIPTION
**Pull request by rakhmets**
## What
Added AM replay protocol (short\single bcopy\single zcopy\multi bcopy) for new infrastructure.

## Why ?
The changes are part of UCP transition to new infrastructure.

### CODE REVIEW

File: src/ucp/Makefile.am [hunk 1]
```diff
@@ -161,6 +163,4 @@ libucp_la_SOURCES = \
 	wireup/wireup.c \
 	wireup/wireup_cm.c \
 	stream/stream_send.c \
-	stream/stream_recv.c \
-	am/eager_single.c \
```
Comment (brminich) [inline src/ucp/Makefile.am:165]: if it's moved to keep alphabetical order, maybe it's worth moving `stream*` as well?

File: src/ucp/am/eager_single.c [hunk 2]
```diff
@@ -17,16 +17,18 @@
 #include <ucp/proto/proto_am.inl>
 
 
-static ucs_status_t ucp_eager_short_progress(uct_pending_req_t *self)
+static ucs_status_t
+ucp_eager_short_commmon_progress(uct_pending_req_t *self, int is_reply)
```
Comment (brminich) [inline src/ucp/am/eager_single.c:21]: make it inline

File: src/ucp/am/eager_single.c [hunk 3]
```diff
@@ -100,12 +128,37 @@ ucp_proto_t ucp_am_eager_short_proto = {
...
+    .desc     = UCP_PROTO_SHORT_DESC,
+    .flags    = 0,
+    .init     = ucp_proto_eager_short_reply_init,
+    .query    = ucp_proto_single_query,
+    .progress = {ucp_eager_short_reply_progress},
+    .abort    = ucp_proto_request_bcopy_abort
+};
+
+static size_t
+ucp_am_eager_single_bcopy_common_pack(void *dest, void *arg, int is_reply)
```
Comment (brminich) [inline src/ucp/am/eager_single.c:155]: make it inline

File: src/ucp/am/eager_single.c [hunk 4]
```diff
@@ -287,28 +381,76 @@ static ucs_status_t ucp_proto_request_eager_am_zcopy_single_init(
...
+    .name     = "am/egr/single/zcopy",
+    .desc     = UCP_PROTO_ZCOPY_DESC,
+    .flags    = 0,
+    .init     = ucp_proto_am_eager_single_zcopy_init,
+    .query    = ucp_proto_single_query,
+    .progress = {ucp_proto_am_eager_single_zcopy_progress},
+    .abort    = ucp_proto_request_bcopy_abort
+};
+
+static ucs_status_t ucp_proto_am_eager_single_zcopy_replay_send_func(
```
Comment (brminich) [inline src/ucp/am/eager_single.c:410]: `replay` -> `reply`
please replace all occurrences

**Comment by rakhmets** [comment]
> we need to add new `ucp_operation_id_t`, like `UCP_OP_ID_AM_SEND` and check it for the corresponding protocols (like it's done for `UCP_OP_ID_TAG_SEND` and `UCP_OP_ID_TAG_SEND_SYNC`)

Done

File: test/gtest/ucp/test_ucp_am.cc [hunk 5]
```diff
@@ -958,11 +958,18 @@ class test_ucp_am_nbx_eager_data_release : public test_ucp_am_nbx {
         m_data_ptr = NULL;
     }
 
-    static void get_test_variants(std::vector<ucp_test_variant> &variants)
+    static void
+    get_test_variants_replay(std::vector<ucp_test_variant> &variants)
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_am.cc:962]: `replay` -> `reply`

File: test/gtest/ucp/test_ucp_am.cc [hunk 6]
```diff
@@ -958,11 +958,18 @@ class test_ucp_am_nbx_eager_data_release : public test_ucp_am_nbx {
...
-        add_variant_values(variants, test_ucp_am_base::get_test_variants, 1,
-                           "proto");
+        add_variant_values(variants, test_ucp_am_base::get_test_variants,
+                           UCP_AM_SEND_FLAG_REPLY, "reply");
+    }
+
+    static void get_test_variants(std::vector<ucp_test_variant> &variants)
+    {
+        add_variant_values(variants, get_test_variants_replay, 0);
+        add_variant_values(variants, get_test_variants_replay, 1, "proto");
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_am.cc:972]: `replay` -> `reply`

**Review by brminich — APPROVED** [review]

File: src/ucp/am/eager_single.c [hunk 7]
```diff
@@ -227,7 +341,7 @@ static void ucp_proto_eager_am_zcopy_add_footer(ucp_request_t *req,
 }
 
 static ucs_status_t
-ucp_proto_eager_am_zcopy_single_send_func(ucp_request_t *req,
+ucp_proto_am_eager_single_zcopy_send_func(ucp_request_t *req,
```
Comment (yosefe) [inline src/ucp/am/eager_single.c:344]: IMO for consistency the functions in this file should start with ucp_am_... (but if it's too big change can be in another PR)

Comment (rakhmets) [reply src/ucp/am/eager_single.c:350]: Ok. I will open a separate PR to update names right after merging https://github.com/openucx/ucx/pull/7988 and https://github.com/openucx/ucx/pull/8019.

**Review by yosefe — APPROVED** [review]

**Review by brminich — APPROVED** [review]

**Review by brminich — APPROVED** [review]

**Review by yosefe — APPROVED** [review]
pls squash

**Review by yosefe — APPROVED** [review]
